Synchronous progress reporting in tests

pepeiborra opened the pull request:
The WorkDoneBegin and WorkDoneEnd notifications are now sent synchronously for tests, to avoid timing issues reported in #1749

ndmitchell commented on ghcide/src/Development/IDE/Core/Shake.hs:
```diff
@@ -183,8 +183,8 @@ data ShakeExtras = ShakeExtras
-    -- ^ How many rules are running for each file
+    ,inProgress :: forall a . NormalizedFilePath -> Action a -> Action a
+    -- ^ Report progress for a rule
     ,progressUpdate :: ProgressEvent -> IO ()
```
Did you consider just putting a whole ProgressReporting in ShakeExtras, so you can keep it a bit more abstract?

ndmitchell commented on ghcide/src/Development/IDE/Core/ProgressReporting.hs:
```diff
@@ -0,0 +1,229 @@
+{-# LANGUAGE RankNTypes #-}
+module Development.IDE.Core.ProgressReporting
```
Great call to pull this out as a module

ndmitchell commented on ghcide/src/Development/IDE/Core/ProgressReporting.hs:
```diff
@@ -0,0 +1,229 @@
+  }
+
+-- | A 'ProgressReporting' that sends the WorkDone Begin and End notifications
+--   synchronously. Progress notifications are sent from a sampling thread.
```
Perhaps a line of comment saying why we want this (namely testing), and the guarantee it providers (you will always get the messages)

ndmitchell commented on ghcide/src/Development/IDE/Core/ProgressReporting.hs:
```diff
@@ -0,0 +1,229 @@
+        f file shift = atomicModifyIORef'_ inProgressVar $
+                HMap.insertWith (\_ x -> shift x) file (shift 0)
+
+        progressLoop :: Double -> LSP.LspM a ()
```
Maybe use `Seconds` from the extra library here?

ndmitchell commented on ghcide/src/Development/IDE/Core/ProgressReporting.hs:
```diff
@@ -0,0 +1,229 @@
+                    current <- liftIO $ readIORef inProgressVar
+                    prev <- progress style prev current t
+                    liftIO $ sleep sample
+                    progressLoop prev
```
It might be clearer if you calculated the value inside the case, then had the `sleep` and `processLoop` at the end of the case, so its easier to see the control flow is just an uninterupted loop.

ndmitchell commented on ghcide/src/Development/IDE/Core/ProgressReporting.hs:
```diff
@@ -0,0 +1,229 @@
+    --   2. KickStarted - transitions from Idle into Reporting
+    progressThread mostRecentProgressEvent inProgress = progressLoopIdle
+      where
+        progressLoopIdle = do
```
I think this does what you are describing, but it's really hard to see with the retries and the running of the reporter async with a kill. Is there a more direct encoding? Like when we get a `progressUpdate Started` we sleep, then spawn the thread, and when we get `Completed` we kill it? I spent a long time trying to find the loop, before realising the loop isn't even in `progressLoop`, it's in `lspShakeProgress`.

pepeiborra replied:
I'm confused too. There's actually two loops, outer one in the mutually recursive functions forming the state machine, and inner one in `lspShakeProgress`. 

I think the code works fine and I don't see how to make it less confusing, so I've just added more comments.

ndmitchell replied:
I was expecting the code to read something along the lines of:

```
state <- newVar Nothing
let progressUpdate KickCompleted =
        join $ modifyVar state $ \case
             Nothing -> pure (Nothing, pure ())
             Just thread -> pure (Nothing, cancel thread)
    progressUpdate KickStarted =
        modifyVar_ state $ \case
             Nothing -> Just <$> async (mRunLspT ...)
             Just thread -> pure $ Just thread
```

So you don't need the outer loop. You'd also make progressStop be the same as KickCompleted, simplifying things further.

pepeiborra replied:
That code is pretty much equivalent to `directProgressReporting` introduced in this PR for tests. And this is a good things, because in hindsight it seems a terrible idea to have one implementation for testing and a different one for production. So the path is clear - just use `directProgressReporting` for both and remove the old version.

wz1000 commented on ghcide/src/Development/IDE/Core/ProgressReporting.hs:
```diff
@@ -0,0 +1,228 @@
+        u <- liftIO newProgressToken
+
+        void $ LSP.sendRequest LSP.SWindowWorkDoneProgressCreate
+            LSP.WorkDoneProgressCreateParams { _token = u } $ const (pure ())
```
According to the spec, we should wait for the response to this request before proceeding. This could be done by using `signalBarrier` instead of `const (pure ())` as the response callback.

pepeiborra commented:
Thank you both for the careful review.

pepeiborra commented:
This PR introduces a couple of new bugs: 

1. I moved the sleep call in `lspShakeProgress` into the `loop`, so now it no longer delays the Begin message and it happens unconditionally. I believe our tests will not catch this.
2. I added a barrier to wait for the Begin response, but it doesn't look like lsp-test sends this response. It breaks a few tests, and if we use this in prod it could make the IDE less responsive unless done in a background thread.

Having one `ProgressReporting` for testing and another one for production is not really a good idea. I'll need to take another look, so moving to draft for now.

wz1000 commented:
> I added a barrier to wait for the Begin response, but it doesn't look like lsp-test sends this response. It breaks a few tests, and if we use this in prod it could make the IDE less responsive unless done in a background thread.

I'm pretty sure I fixed this in https://github.com/haskell/lsp/commit/5fa33f89d06a5623a4fd06fd2c0fdd4dbbc9c32d#diff-8e90cf08428164fe5d7ef53f4d8e7900169c97415b3198f3040e75b96803e009

pepeiborra commented:
> > I added a barrier to wait for the Begin response, but it doesn't look like lsp-test sends this response. It breaks a few tests, and if we use this in prod it could make the IDE less responsive unless done in a background thread.
> 
> I'm pretty sure I fixed this in [haskell/lsp@5fa33f8#diff-8e90cf08428164fe5d7ef53f4d8e7900169c97415b3198f3040e75b96803e009](https://github.com/haskell/lsp/commit/5fa33f89d06a5623a4fd06fd2c0fdd4dbbc9c32d#diff-8e90cf08428164fe5d7ef53f4d8e7900169c97415b3198f3040e75b96803e009)

My bad, it's not lsp-test fault - it was a division by zero bug in the `progress` loop when `todo` is 0. This was likely the real issue identified in #1749 - I'll send a minimal fix in a separate PR

pepeiborra commented:
Replaced by #1770
